### ISSUE TITLE
Build error with version 2.3.1: @parcel/core: Failed to resolve 'process' from ...

### ISSUE THREAD

**Issue by spointecker** [parcel-bundler/parcel#7697]
# 🐛 bug report

I just upgraded to version 2.3.1 and now the build process fails.
Don't know exactly if this is a parcel or a sentry issue?

## 🎛 Configuration (.babelrc, package.json, cli command)

```js
{
    "presets": [
        [
            "@babel/preset-env",
            {
                "modules": false,
                "useBuiltIns": false,
                "targets": {
                    "node": "current"
                }
            }
        ],
        "@babel/preset-react"
    ],
    "plugins": [],
    "env": {
        "test": {
            "plugins": ["@babel/plugin-transform-modules-commonjs"]
        }
    }
}

```

## 🤔 Expected Behavior

Build should pass.

## 😯 Current Behavior

The build fails with the following exception:
```
🚨 Build failed.

@parcel/core: Failed to resolve 'process' from './node_modules/@sentry/utils/esm/node.js'

  /Users/*****/Sites/*****/node_modules/@sentry/utils/esm/node.js:18:43
    17 | /**
  > 18 |  * Requires a module which is protected against bundler minification.
  >    |                                           ^^^^^^^^
    19 |  *
    20 |  * @param request The module path to resolve
```

## 💁 Possible Solution

idk.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.3.1
| Node             | 16.11.1
| npm/Yarn         | yarn: 1.22.17
| Operating System | mac osx

**Comment by klausbreyer** [comment]
Having the same issue. 
For me it is with the contentful npm module.



```
rm -rf .parcel-cache && yarn start
yarn run v1.22.15
warning package.json: No license field
$ parcel public/index.html
Server running at http://localhost:1234
🚨 Build failed.

@parcel/core: Failed to resolve 'process' from
'./node_modules/contentful-sdk-core/dist/index.es-modules.js'

  /Users/kb1/versioned/yhb113-fillter.me-unified/web/node_modules/contentful-sdk-core/dist/index.es-modules.js:257:17
    256 |    **/
  > 257 | peof process !== 'undefined' && !process.browser;
  >     |      ^^^^^^^
    258 | }
    259 | function isReactNative() {
```

It started with **2.3.0.** 

.parcelrc

```{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.css": [
      "@parcel/transformer-css-experimental"
    ]
  },
  "optimizers": {
    "*.css": [
      "@parcel/optimizer-css"
    ]
  }
}
```

(Same behaviour with empty .parcelrc)

**Comment by koskokos2** [comment]
Have the same problem with `axios`.

Temporary workaround is to add alias to `package.json` :
```
"alias": {
    "process": false
 }
```

**Comment by mattlockyer** [comment]
I was having this issue related to `js-sha256`

https://github.com/parcel-bundler/parcel/issues/7060#issuecomment-1035880313

This solved it, in builds and dev. Basically, define process for the browser as an empty object:
```
"alias": {
		"process": {
			"global": "{}"
		}
	},
	```

**Comment by devongovett** [comment]
I tried the following code:

```js
import axios from 'axios';
console.log(axios)
```

and the console output was:

```
Building...

@parcel/resolver-default: Auto installing polyfill for Node builtin module "process"...

  /Users/devongovett/Downloads/test-92/node_modules/axios/lib/defaults.js:22:21
    21 |     adapter = require('./adapters/xhr');
  > 22 |   } else if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
  >    |                     ^^^^^^^ used here
    23 |     // For node use HTTP adapter
    24 |     adapter = require('./adapters/http');

  📝 Learn more: https://parceljs.org/features/node-emulation/#polyfilling-%26-excluding-builtin-node-modules

Installing process...
Bundling...
Packaging & Optimizing...
✨ Built in 1.48s

dist/test.js    19.37 KB    286ms
```

So, it auto installed `process` successfully and built. Is there another way to reproduce this?

**Comment by koskokos2** [comment]
@devongovett for me it is reproducible quite easily. The simplest example would be 2 file project:

index.html:
```
<script type="module">
  import axios from "axios";
</script>

```

package.json:
```
  "scripts": {
    "start": "parcel index.html"
  },
  "devDependencies": {
    "parcel": "2.3.1"
  },
  "dependencies": {
    "axios": "0.25.0"
  }
```


that results in the error noted by others above:
```
npm start

> @ start /Users/kos/test
> parcel index.html

Server running at http://localhost:1234
🚨 Build failed.

@parcel/core: Failed to resolve 'process' from './node_modules/axios/lib/defaults.js'

  /Users/kostiantyn.moroz/Documents/parcel/packages/examples/react-refresh/node_modules/axios/lib/defaults.js:22:21
    21 |     adapter = require('./adapters/xhr');
  > 22 |   } else if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
  >    |                     ^^^^^^^
    23 |     // For node use HTTP adapter
    24 |     adapter = require('./adapters/http');
```

Tried on both Node 14 and Node 16 with the same outcome. 

I also tried enabling more logging by `npm start -- --log-level verbose` but as soon as it encounters any error it replaces any verbose output with just this error resulting in verbose logging appearing on the screen for a split second.

**Comment by chang-ryan** [comment]
just linking my issue repro here as well: https://github.com/parcel-bundler/parcel/issues/7060#issuecomment-1035900934

also a problem with `js-sha3` like @mattlockyer 

glad to help contribute to this fix as well

**Comment by devongovett** [comment]
@koskokos2 if you run Parcel and pipe the output to cat you can see the full output.

`parcel index.html | cat`

**Comment by koskokos2** [comment]
Oh, thanks. Looks like it tried to auto-install `process` but didn't. With axios supporting both Node and browser it should not really care about polyfills for it. I would probably even prefer an option to disable this auto-install Node polyfills altogether. Here is my output anyway:

```
npm start -- --log-level verbose | cat

> start
> parcel index.html "--log-level" "verbose"

Server running at http://localhost:1234

@parcel/resolver-default: Auto installing polyfill for Node builtin module "process"...

  /Users/kos/test/node_modules/axios/lib/defaults.js:22:21
    21 |     adapter = require('./adapters/xhr');
  > 22 |   } else if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
  >    |                     ^^^^^^^ used here
    23 |     // For node use HTTP adapter
    24 |     adapter = require('./adapters/http');

  📝 Learn more: https://parceljs.org/features/node-emulation/#polyfilling-%26-excluding-builtin-node-modules

🚨 Build failed.

@parcel/core: Failed to resolve 'process' from './node_modules/axios/lib/defaults.js'

  /Users/kos/test/node_modules/axios/lib/defaults.js:22:21
    21 |     adapter = require('./adapters/xhr');
  > 22 |   } else if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
  >    |                     ^^^^^^^
    23 |     // For node use HTTP adapter
    24 |     adapter = require('./adapters/http');
```

**Comment by chang-ryan** [comment]
another weird thing i noticed using the `alias` workaround

build works when explicitly providing a source
```json
this works
...
"scripts": {
  "serve": "parcel src/index.html",
  "build": "parcel build src/index.html"
},
...
```

but it doesn't work when you state a `source` as a field in the `package.json`
```json
this does not work
...
"source": "src/index.html",
"scripts": {
  "serve": "parcel src/index.html",
  "build": "parcel build"
},
...
```

**Comment by CrommVardek** [comment]
@koskokos2 :

For me, I've added the following to my package.json :
`
"alias": {
    "process": false
 }
`

Now I can build on docker and it's working fine on nginx (ya!) but on local it does not work anymore... (Process is undefined) 

I think I'm gonna downgrade to parcel 2.2 for now until it's resolved...

**Comment by henri-hulski** [comment]
Here I've the same problem with `overmind` and `overmind-react`.
They're just using `process` to check if it's a node environment.
So the `"alias": { "process": false }` hack worked here.

**Comment by niicojs** [comment]
Same issue, also with Axios.
Parcel doesn't seeem to auto install process polyfill.

Update: I can see it briefly that it detects there is `process` to be installed but it quickly get back to the error.

**Comment by ilukin89** [comment]
Adding `"alias": { "process": false }` worked for me too. Thank you everyone, I had issues for days with axios error

**Comment by tpudel** [comment]
Having this error as well from migration from v1 to v2, even downgrading to 2.2 did not work, polyfill does not seem to be installing like the logs say.

`"alias": { "process": "process/browser.js", "assert": "assert" }` worked for me to get around this, but had to manually install both `process` and `assert` before the aliases were utilized.

Alias Reference: https://parceljs.org/features/node-emulation/

**Comment by nwhitmont** [comment]
I'm upgrading from 2.2.1 to 2.3.1, this time I'm getting same issue with NodeJS built-ins, for me it is Buffer.

`@parcel/core: Failed to resolve 'buffer'`

- Node: LTS 16.14.0
- Yarn: 1.22.15
- OS: Windows (local dev), Linux (Bitbucket pipeline)

**Comment by ekeyser** [comment]
For folks seeing issues with `process` and `buffer`, use @henri-hulski solution above and also node v17. Node v17 resolved issue with `buffer` for me.

**Comment by jbcz-tv** [comment]
I was having this issue related to `debug`

**Comment by vladinator1000** [comment]
I get a similar error when trying to build Apollo Client, where [`ts-invariant`](https://github.com/apollographql/invariant-packages/tree/main/packages/ts-invariant) fails to build with a `@parcel/core: Failed to resolve 'process` message.

Similarly to @tpudel I 
`yarn add proecess`
then added this to package.json
```
"alias": { "process": "process/browser.js" }
```

**Comment by humphd** [comment]
I'm hitting this when using [aws-amplify](https://www.npmjs.com/package/aws-amplify):

```
parcel index.html | cat
Server running at http://localhost:1234
Building...

@parcel/resolver-default: Auto installing polyfill for Node builtin module
"process"...

  /private/tmp/fragments-ui/node_modules/@aws-amplify/core/src/JS.ts:178:10
    177 |   const isNode =
  > 178 |     typeof process !== 'undefined' &&
  >     |            ^^^^^^^^ used here
    179 |     process.versions != null &&
    180 |     process.versions.node != null;

  📝 Learn more:
     https://parceljs.org/features/node-emulation/#polyfilling-%26-excluding-builtin-node-modules

🚨 Build failed.

@parcel/core: Failed to resolve 'process' from
'./node_modules/@aws-amplify/core/lib-esm/JS.js'

  /private/tmp/fragments-ui/node_modules/@aws-amplify/core/lib-esm/JS.js:178:10
    177 |     }
  > 178 |     return ret;
  >     |          ^^^^^^^
    179 | };
    180 | /**
```

I tried to reduce it, but can't make it happen in isolation.

**Comment by rajsite** [comment]
Ran into a similar sounding error using 2.3.1:
```
@parcel/core: Failed to resolve 'buffer' from './src/main.ts'
```

I had to go to the [Node Emulation: Polyfilling & Excluding Builtin Node Modules
](https://parceljs.org/features/node-emulation/#polyfilling-%26-excluding-builtin-node-modules) table, find the package used for buffer, ie [buffer](https://www.npmjs.com/package/buffer), and manually install it. After that parcel was able to build successfully.

Similar to what tpudel described: https://github.com/parcel-bundler/parcel/issues/7697#issuecomment-1040867550 but I did not need to create the alias for buffer, just install the npm package.

node v14.17.1 npm 7.24.2 parcel 2.3.1

**Comment by MuhammadM1998** [comment]
I'm having the same issue with firebase firestore
```
@parcel/core: Failed to resolve 'process' from './node_modules/@firebase/firestore/dist/index.esm2017.js'

  D:\Workspace\Front-End\Apps\RISC-Aswan\node_modules\@firebase\firestore\dist\index.esm2017.js:5853:1
    5852 |         return t.store(e);
  > 5853 |     }
  >      | ^
    5854 |     // visible for testing
    5855 |     /** Parse User Agent to determine iOS version. Returns -1 if not found. */
```

Tried the mentioned solutions but nothing worked for me, here's what I did:
- used alias in my package.json file
```
 "alias": {
        "process": "false"
    }
```
- manually installed process package
 ```
"dependencies": {
        "process": "^0.11.10",
    },
```
- updated node to v16.14.0 instead of v16.13.1.
- used parcel build ./src/index.html and removed "source": "./src/index.html" in the package.json

Here's my package.json file snippet:
```
"scripts": {
        "dev": "parcel ./src/index.html --https --open firefox",
        "build": "npm run buildRelative && npm run updatePaths",
        "buildRelative": "parcel build --no-source-maps --public-url .",
        "updatePaths": "(Get-Content -Raw public/index.html) -replace 'src=\"', 'src=\"./' | Set-Content -NoNewLine public/index.html"
    },
    "devDependencies": {
        "autoprefixer": "^10.4.2",
        "parcel": "^2.3.1",
        "postcss": "^8.4.6",
        "tailwindcss": "^3.0.18"
    },
    "dependencies": {
        "firebase": "^9.6.6",
        "process": "^0.11.10",
        "vanilla-hamburger": "^0.2.3"
    },
    "alias": {
        "process": "false"
    }
```

**Comment by MuhammadM1998** [comment]
> I'm having the same issue with firebase firestore
> 
> ```
> @parcel/core: Failed to resolve 'process' from './node_modules/@firebase/firestore/dist/index.esm2017.js'
> 
>   D:\Workspace\Front-End\Apps\RISC-Aswan\node_modules\@firebase\firestore\dist\index.esm2017.js:5853:1
>     5852 |         return t.store(e);
>   > 5853 |     }
>   >      | ^
>     5854 |     // visible for testing
>     5855 |     /** Parse User Agent to determine iOS version. Returns -1 if not found. */
> ```
> 
> Tried the mentioned solutions but nothing worked for me, here's what I did:
> 
>     * used alias in my package.json file
> 
> 
> ```
>  "alias": {
>         "process": "false"
>     }
> ```
> 
>     * manually installed process package
> 
> 
> ```
> "dependencies": {
>        "process": "^0.11.10",
>    },
> ```
> 
>     * updated node to v16.14.0 instead of v16.13.1.
> 
>     * used parcel build ./src/index.html and removed "source": "./src/index.html" in the package.json
> 
> 
> Here's my package.json file snippet:
> 
> ```
> "scripts": {
>         "dev": "parcel ./src/index.html --https --open firefox",
>         "build": "npm run buildRelative && npm run updatePaths",
>         "buildRelative": "parcel build --no-source-maps --public-url .",
>         "updatePaths": "(Get-Content -Raw public/index.html) -replace 'src=\"', 'src=\"./' | Set-Content -NoNewLine public/index.html"
>     },
>     "devDependencies": {
>         "autoprefixer": "^10.4.2",
>         "parcel": "^2.3.1",
>         "postcss": "^8.4.6",
>         "tailwindcss": "^3.0.18"
>     },
>     "dependencies": {
>         "firebase": "^9.6.6",
>         "process": "^0.11.10",
>         "vanilla-hamburger": "^0.2.3"
>     },
>     "alias": {
>         "process": "false"
>     }
> ```

for some reason modifying alias to the following worked in dev and build
```
"alias": {
    "process": {
        "global": "{}"
    }
}
```

**Comment by jamiemcconnell** [comment]
> ```"alias": {
> "process": {
>        "global": "{}"
>    }
> }
Just changing process to the above as @MuhammadM1998 has _without_ installing `process` as a dependancy works

**Comment by MuhammadM1998** [comment]
> > ```
> > "process": {
> >        "global": "{}"
> >    }
> > }
> > ```
> 
> Just changing process to the above as @MuhammadM1998 has _without_ installing `process` as a dependancy works

I removed process as a dependancy and it works well for me too, yet i'm waiting for a fixed version to stop using such workarounds.

**Comment by eightyfive** [comment]
It does _not_ work. All `process.env.VARIABLE_NAME` are `undefined`. It just prevents the build to fail, but you can't use the parcel [env variables](https://en.parceljs.org/env.html) feature.

Or am I missing something?

(Tried all workarounds here - they don't bring back the env variables feature)

**Comment by jtojnar** [comment]
Did you try

```
"process": {
    "global": "process"
}
```

?

**Comment by humphd** [comment]
@devongovett #7735 amazing, thank you for fixing!

**Comment by eightyfive** [comment]
@jtojnar not this one, will try, ty.

**Comment by devongovett** [comment]
v2.3.2 has a fix. Please let me know if it works for you.

**Comment by humphd** [comment]
> I'm hitting this when using [aws-amplify](https://www.npmjs.com/package/aws-amplify):
> 
> @parcel/resolver-default: Auto installing polyfill for Node builtin module
> "process"...

Confirming that parcel@2.3.2 fixes this error for my case.

**Comment by eightyfive** [comment]
@devongovett Hmmm the build error is gone but my env variables (in `.env` file) are still not loaded (`undefined`)... 😞 😞 

Does someone else experience the same error ?

**Comment by ekeyser** [comment]
@eightyfive hey there - I'm running v14 node, just installed 2.3.2 parcel and I am loading a .env which appear to be loaded correctly. But I'm not sure if there's something with my app/env that's wildly different from yours - I'm sure there is just not sure what info you would need to help with that.

**Comment by eightyfive** [comment]
Then it must come from my setup. Thanks for letting me know.

**Comment by eightyfive** [comment]
Ok, just for following up:

- I had my env file in `<project_root>/src/.env`
- Need to be in `<project_root>/.env`
- I deleted the `.parcel-cache` directory, ran `parcel build`, then re-ran `parcel serve` (somehow needed to run `build` before...)
- All good !

**Comment by KaKi87** [comment]
Hello,

I am experiencing this error :

```
[Error: ENOENT: no such file or directory, open 'node_modules/@sentry/vue/node_modules/@sentry/utils/src/node.ts'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: 'node_modules/@sentry/vue/node_modules/@sentry/utils/src/node.ts'
}
```

It doesn't match any of those reported in this issue, but the fix (adding `"alias": { "process": false }` to `package.json`) works on it.

So, I expected it to be fixed since v2.3.2, but it isn't.
I deleted `.parcel-cache` and `node_modules` at the second attempt, just in case.

Thanks

**Comment by spointecker** [comment]
I upgraded now to `2.3.2` and I experience the same issue as @KaKi87.

**Comment by devongovett** [comment]
Please open a separate issue if you are experiencing something different from the original one.

**Comment by KaKi87** [comment]
Well, I'm not sure if it is something different, which is why I asked here.

**Comment by renjithspace** [comment]
Thanks @devongovett

It working. I was using `v2.3.1` and faced this issue. I couldn't use my `dotenv` variables by accessing like `process.env.APP_URL`. I was receiving build error `@parcel/core: Failed to resolve 'process'`.

After your fix, now I update it to `v2.3.2` and it working as expected.

**Comment by tjx666** [comment]
![image](https://user-images.githubusercontent.com/41773861/180780222-8372c29c-4160-4edd-91be-67c816ac8e3e.png)

Using alias false doesn't work:

```json
{
    "license": "UNLICENSED",
    "alias": {
        "fs": false,
        "os": false,
        "path": false,
        "util": false,Ï
        "process": false
    },
}
```

**Comment by domenkozar** [comment]
Using 2.7.0 I still had to apply the alias workaround.

**Comment by BraffZackland** [comment]
This issue is still present. The alias workaround worked to unblock me.

**Comment by Rigidity** [comment]
Any updates on this? Having the same issue.

**Comment by guoyunhe** [comment]
Still have this error in parcel 2.7.0, with pnpm. Can you fix it?

**Comment by chenzx** [comment]
I've the similar problem. But i use my own `globalThis.process=...` runtime mock in my app code, so i think Parcel should NOT find polyfill when bundling with `typeof process` code...

**Comment by lzm0x219** [comment]
Any updates on this? Having the same issue.

**Comment by njleonzhang** [comment]
still having thi issue on node 14, parce 2.8 with pnpm, no issue with npm

**Comment by JonathonDunford** [comment]
Bump, this is still an issue on 2.8.0
My issue specifically came about when trying to use bcrypt.
Parcel seems to try to polyfill a ton of dev dependencies including AWS SDK and process.

**Comment by wesbos** [comment]
Having is issue with parcel 2.8.2 and the axios library. Anyone know a fix?

**Comment by wesbos** [comment]
The fixes are conflicting now...

Sentry+Utils can be fixed with:

```
  "alias": {
    "process": {
      "global": {}
    }
  },
```

and axios can be fixed with 

```
  "alias": {
    "process": false
  },
```

but how do I fix both?

**Comment by mischnic** [comment]
@wesbos Can you share an example with the sentry and axios versions you're using? And what exactly are you importing from them?

**Comment by wesbos** [comment]
@mischnic Here is a very small reproduction:

https://github.com/wesbos/parcel-issues

**Comment by wesbos** [comment]
Thanks to @mischnic  - found the issue. Apparently [pnpm autoinstall is broken](https://github.com/parcel-bundler/parcel/issues/8304).

Add these dev deps to your package.json. Note I was using parcel in a workspace, but these need to be installed in the project root.

```json
"buffer": "^5.5.0",
"process": "^0.11.10",
```

Then add these alias to package.json:

```json
  "alias": {
    "process": "process/browser.js",
    "buffer": "buffer"
  },
```

Then install and run parcel, and you are back working! 

also see [node emulation](https://parceljs.org/features/node-emulation/#polyfilling-%26-excluding-builtin-node-modules)

**Comment by damian-pastorini** [comment]
The "alias" solution doesn't work for me :( my issue is with "process" on a library I'm requiring.

**Comment by susanbakhtar** [comment]
hello, i did some process for run build. it works. but i didn't see any change in my project file . can i fix this?

<img width="302" alt="Screenshot 1402-08-22 at 05 55 38" src="https://github.com/parcel-bundler/parcel/assets/148472264/aca491da-8b4a-4de5-b3a6-82ccda9f7d4c">

**Comment by skullface** [comment]
I’ve just ran into this with NPM on a tiny project with no dependencies other than Parcel. I imagine my issue might be from not properly closing the port, as my terminal window was hanging so I force-quit it. Reloading my terminal didn’t fix the problem nor did a system restart. I use PNPM for other projects, so that may be the culprit? Wes’ workaround at https://github.com/parcel-bundler/parcel/issues/7697#issuecomment-1360074613 solved the issue (server successfully started).

**Comment by jhobz** [comment]
I just ran into this issue on v2.11.0 in a GitHub action, but not when building locally. From the looks of things, it appears this was marked as fixed in v2.3.2, but yeah, just started getting the same issue in v2.11.0 after installing a new package (that presumably uses `process`).
![image](https://github.com/parcel-bundler/parcel/assets/1765799/74bc2cdb-4f92-45ee-9bf5-5aa7e4d098f0)

Edit: to note, the workaround at [#7697 (comment)](https://github.com/parcel-bundler/parcel/issues/7697#issuecomment-1360074613) does indeed work for me as well.
